services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerFactory.java

```diff
@@ -0,0 +1,430 @@
+      adjustHealthCheck();
+    }
+
+    private boolean isRetryTimerPending() {
+      return retryTimer != null && retryTimer.isPending();
+    }
+
+    // Start or stop health check according to the current states.
+    private void adjustHealthCheck() {
+      if (!disabled && serviceName != null && Objects.equal(rawState.getState(), READY)) {
+        running = true;
+        if (activeRpc == null && !isRetryTimerPending()) {
+          startRpc();
+        }
+      } else {
+        running = false;
+        // Prerequisites for health checking not met.
+        // Make sure it's stopped.
+        stopRpc("Client stops health check");
+        backoffPolicy = null;
+        gotoState(rawState);
+      }
+    }
+
+    private void startRpc() {
+      checkState(activeRpc == null, "previous health-checking RPC has not been cleaned up");
+      checkState(subchannel != null, "init() not called");
+      // Optimization suggested by @markroth: if we are already READY and starting the health
+      // checking RPC, either because health check is just enabled or has switched to a new service
+      // name, we don't go to CONNECTING, otherwise there will be artificial delays on RPCs
+      // waiting for the health check to respond.
+      if (!Objects.equal(concludedState.getState(), READY)) {
+        gotoState(ConnectivityStateInfo.forNonError(CONNECTING));
+      }
+      activeRpc = new HcStream();
+    }
+
+    private void stopRpc(String msg) {
+      if (activeRpc != null) {
+        activeRpc.cancel(msg);
+        // Abandon this RPC.  We are not interested in anything from this RPC any more.
+        activeRpc = null;
+      }
+      if (retryTimer != null) {
+        retryTimer.cancel();
+        retryTimer = null;
+      }
+    }
+
+    private void gotoState(ConnectivityStateInfo newState) {
+      checkState(subchannel != null, "init() not called");
+      if (!Objects.equal(concludedState, newState)) {
+        concludedState = newState;
+        delegate.handleSubchannelState(subchannel, concludedState);
+      }
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("running", running)
+          .add("disabled", disabled)
+          .add("activeRpc", activeRpc)
+          .add("serviceName", serviceName)
+          .add("rawState", rawState)
+          .add("concludedState", concludedState)
+          .toString();
+    }
+
+    private class HcStream extends ClientCall.Listener<HealthCheckResponse> {
+      private final ClientCall<HealthCheckRequest, HealthCheckResponse> call;
+      private String callServiceName;
+      private long callStartNanos;
+      private boolean callHasResponded;
+
+      HcStream() {
+        callStartNanos = time.currentTimeNanos();
+        call = subchannel.asChannel().newCall(HealthGrpc.getWatchMethod(), CallOptions.DEFAULT);
+        callServiceName = serviceName;
+        call.start(this, new Metadata());
```

Comment:
Could we please not issue the RPC in the constructor? It's fine to create+assign the `call`, but let's move the start to a separate method to make it more clear at the caller.

Reply:
Done.